File: src/zoneA/doorTarget.ts
import Phaser from 'phaser';
import type { BallBodyData } from '../core/contracts';
import * as Layout from '../core/Layout';
import { nearestDoorBall } from './ballMath';

/**
 * A Matter body carrying a Zone A ball's identity, as seen through the shared physics world.
 * `circleRadius` (Matter sets it on circle bodies) measures edge distance; `gameObject` is
 * Phaser's back-reference to the Matter.Image so Zone C can remove the ball by destroying it.
 */
export interface DoorBall {
  position: { x: number; y: number };
  circleRadius?: number;
  gameObject?: Phaser.GameObjects.GameObject;
  ballData?: BallBodyData;
}

/**
 * The Zone A ball a trap-door tap would grab: nearest the fixed door mouth
 * (`Layout.zoneBEntry.x` / `Layout.zoneC.y`) by edge distance, read from the SHARED Matter
 * world. This is the single source of truth for BOTH Zone C's actual grab (`onTap`) and Zone
 * A's candidate highlight, so the ring can never sit on a different ball than the one the tap
 * takes. The pure selection math lives in `nearestDoorBall` (unit-tested); this only gathers
 * the world's bodies and hands over the door geometry.
 */
export function findNearestDoorBall(scene: Phaser.Scene): DoorBall | undefined {
  const bodies = scene.matter.world.getAllBodies() as unknown as DoorBall[];
  return nearestDoorBall(bodies, Layout.zoneBEntry.x, Layout.zoneC.y);
}

File: Assets/Game/Gameplay/DoorTarget.cs
using System.Collections.Generic;
using RichCoast.Core;

namespace RichCoast.Game
{
    /// <summary>
    /// The Zone A ball a trap-door tap would grab — nearest the door mouth by edge distance, read
    /// straight off the live <see cref="Board"/>. The single source of truth for BOTH Zone C's actual
    /// grab and Zone A's candidate glow, so the glow can never sit on a different ball than the one
    /// the tap takes. The selection math is the pure, unit-tested <see cref="DoorMath"/>.
    /// </summary>
    public static class DoorTarget
    {
        static readonly List<DoorCandidate> scratch = new List<DoorCandidate>();
        static readonly List<Ball> order = new List<Ball>();

        /// <summary>The door mouth sits at the funnel apex — the world origin by construction.</summary>
        const double MouthX = 0, DoorY = 0;

        public static Ball Find(Board board)
        {
            scratch.Clear();
            order.Clear();
            foreach (var ball in board.Balls)
            {
                if (ball.Consumed) continue; // mid-merge: it's about to vanish
                var p = ball.Position;
                scratch.Add(new DoorCandidate(p.x, p.y, ball.Radius));
                order.Add(ball);
            }
            int i = DoorMath.NearestDoorBall(scratch, MouthX, DoorY);
            return i < 0 ? null : order[i];
        }
    }
}
